<a href="https://colab.research.google.com/github/JohnMakgakgaDS/ML-intern-Cognifyz/blob/main/mltask1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
#Load data from the local machine into colab
from google.colab import files

uploaded = files.upload()

Saving Dataset.csv to Dataset.csv


In [ ]:
#Libraries or data science packages to be used along the program tasks

import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.metrics import mean_squared_error, r2_score


In [ ]:
# Assuming the uploaded file is in CSV format, load it into a pandas DataFrame
# Replace 'your_file.csv' with the actual file name you uploaded
file_name = list(uploaded.keys())[0]  # Get the uploaded file name
df = pd.read_csv(file_name)

# View the first few rows of the dataset
df.head()

Restaurant ID         Restaurant Name  Country Code              City  \
0        6317637        Le Petit Souffle           162       Makati City   
1        6304287        Izakaya Kikufuji           162       Makati City   
2        6300002  Heat - Edsa Shangri-La           162  Mandaluyong City   
3        6318506                    Ooma           162  Mandaluyong City   
4        6314302             Sambo Kojin           162  Mandaluyong City   

                                             Address  \
0  Third Floor, Century City Mall, Kalayaan Avenu...   
1  Little Tokyo, 2277 Chino Roces Avenue, Legaspi...   
2  Edsa Shangri-La, 1 Garden Way, Ortigas, Mandal...   
3  Third Floor, Mega Fashion Hall, SM Megamall, O...   
4  Third Floor, Mega Atrium, SM Megamall, Ortigas...   

                                     Locality  \
0   Century City Mall, Poblacion, Makati City   
1  Little Tokyo, Legaspi Village, Makati City   
2  Edsa Shangri-La, Ortigas, Mandaluyong City   
3      SM Megamall, Ortigas, Mandaluyong City   
4      SM Megamall, Ortigas, Mandaluyong City   

                                    Locality Verbose   Longitude   Latitude  \
0  Century City Mall, Poblacion, Makati City, Mak...  121.027535  14.565443   
1  Little Tokyo, Legaspi Village, Makati City, Ma...  121.014101  14.553708   
2  Edsa Shangri-La, Ortigas, Mandaluyong City, Ma...  121.056831  14.581404   
3  SM Megamall, Ortigas, Mandaluyong City, Mandal...  121.056475  14.585318   
4  SM Megamall, Ortigas, Mandaluyong City, Mandal...  121.057508  14.584450   

                           Cuisines  ...          Currency Has Table booking  \
0        French, Japanese, Desserts  ...  Botswana Pula(P)               Yes   
1                          Japanese  ...  Botswana Pula(P)               Yes   
2  Seafood, Asian, Filipino, Indian  ...  Botswana Pula(P)               Yes   
3                   Japanese, Sushi  ...  Botswana Pula(P)                No   
4                  Japanese, Korean  ...  Botswana Pula(P)               Yes   

  Has Online delivery Is delivering now Switch to order menu Price range  \
0                  No                No                   No           3   
1                  No                No                   No           3   
2                  No                No                   No           4   
3                  No                No                   No           4   
4                  No                No                   No           4   

   Aggregate rating  Rating color Rating text Votes  
0               4.8    Dark Green   Excellent   314  
1               4.5    Dark Green   Excellent   591  
2               4.4         Green   Very Good   270  
3               4.9    Dark Green   Excellent   365  
4               4.8    Dark Green   Excellent   229  

[5 rows x 21 columns]

#Task 1 : Predict Restaurant Ratings

#Objective: Build a machine learning model to predict the aggregate rating of a
#restaurant based on other features.

Preprocess the dataset by handling missing values,
encoding categorical variables, and splitting the data
into training and testing sets.

#show number of rows and columns

In [ ]:
df.shape

(9551, 21)

There are 9551 rows in the dataset
There are 21 columns in the dataset

#check missing values in each column

In [ ]:
df.isnull().sum()

Restaurant ID           0
Restaurant Name         0
Country Code            0
City                    0
Address                 0
Locality                0
Locality Verbose        0
Longitude               0
Latitude                0
Cuisines                9
Average Cost for two    0
Currency                0
Has Table booking       0
Has Online delivery     0
Is delivering now       0
Switch to order menu    0
Price range             0
Aggregate rating        0
Rating color            0
Rating text             0
Votes                   0
dtype: int64

There's 9 missing values in column Cuisines

Handling missing values for Cuisines

In [ ]:
 #Fill missing values with a placeholder
df['Cuisines'].fillna('Unknown', inplace=True)


<ipython-input-9-ea5a34f7acc5>:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['Cuisines'].fillna('Unknown', inplace=True)


In [ ]:
#checking for validation
df.isnull().sum()

Restaurant ID           0
Restaurant Name         0
Country Code            0
City                    0
Address                 0
Locality                0
Locality Verbose        0
Longitude               0
Latitude                0
Cuisines                0
Average Cost for two    0
Currency                0
Has Table booking       0
Has Online delivery     0
Is delivering now       0
Switch to order menu    0
Price range             0
Aggregate rating        0
Rating color            0
Rating text             0
Votes                   0
dtype: int64

Filled missing values with Unknown , There are no missing values .

#Encoding categorical variables

In [ ]:
df['Has Table booking'] = df['Has Table booking'].replace({'Yes': 1, 'No': 0})
df['Has Online delivery'] = df['Has Online delivery'].replace({'Yes': 1, 'No': 0})
df['Is delivering now'] = df['Is delivering now'].replace({'Yes': 1, 'No': 0})
df['Switch to order menu'] = df['Switch to order menu'].replace({'Yes': 1, 'No': 0})
df['Rating text'] = df['Rating text'].replace({'Poor': 0, 'Not rated': 1,'Average':2,'Good':3,'Very Good':4,'Excellent':5 })
df['Rating color'] = df['Rating color'].replace({'Red': 0, 'White': 1,'Orange':2,'Yellow':3,'Green':4,'Dark Green':5 })

<ipython-input-13-b147e7f11672>:1: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df['Has Table booking'] = df['Has Table booking'].replace({'Yes': 1, 'No': 0})
<ipython-input-13-b147e7f11672>:2: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df['Has Online delivery'] = df['Has Online delivery'].replace({'Yes': 1, 'No': 0})
<ipython-input-13-b147e7f11672>:3: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=Fa

In [ ]:
#Validation
df.iloc[0:5,10:21]

Average Cost for two          Currency  Has Table booking  \
0                  1100  Botswana Pula(P)                  1   
1                  1200  Botswana Pula(P)                  1   
2                  4000  Botswana Pula(P)                  1   
3                  1500  Botswana Pula(P)                  0   
4                  1500  Botswana Pula(P)                  1   

   Has Online delivery  Is delivering now  Switch to order menu  Price range  \
0                    0                  0                     0            3   
1                    0                  0                     0            3   
2                    0                  0                     0            4   
3                    0                  0                     0            4   
4                    0                  0                     0            4   

   Aggregate rating  Rating color  Rating text  Votes  
0               4.8             5            5    314  
1               4.5             5            5    591  
2               4.4             4            4    270  
3               4.9             5            5    365  
4               4.8             5            5    229

#Define features and set target variable

In [ ]:
features = ['Has Table booking','Has Online delivery','Is delivering now','Switch to order menu','Price range','Rating color','Rating text']
target = 'Aggregate rating'

#Feature Selection

In [ ]:
X = df[features]
y = df[target]

#Split the data into training and testing sets

In [ ]:
from sklearn.model_selection import train_test_split

X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.2, random_state=20)

#Select a regression algorithm (e.g., linear regression,
#decision tree regression) and train it on the training data.

In [ ]:
# Initialize the Linear Regression model
linear_reg_model = LinearRegression()

# Train the model on the training data
linear_reg_model.fit(X_train, y_train)


LinearRegression()

In [ ]:

# Initialize the Decision Tree Regressor model
decision_tree_model = DecisionTreeRegressor(random_state=42)

# Train the model on the training data
decision_tree_model.fit(X_train, y_train)


DecisionTreeRegressor(random_state=42)

#Make Predictions

In [ ]:
# Make predictions using the Linear Regression model
y_pred_linear = linear_reg_model.predict(X_test)

# Make predictions using the Decision Tree Regressor model
y_pred_tree = decision_tree_model.predict(X_test)


#Evaluate the Models

In [ ]:
# Evaluate the Linear Regression model
mse_linear = mean_squared_error(y_test, y_pred_linear)
r2_linear = r2_score(y_test, y_pred_linear)

# Evaluate the Decision Tree Regression model
mse_tree = mean_squared_error(y_test, y_pred_tree)
r2_tree = r2_score(y_test, y_pred_tree)

# Print out the evaluation results
print(f"Linear Regression MSE: {mse_linear:.2f}")
print(f"Linear Regression R²: {r2_linear:.2f}")
print(f"Decision Tree Regression MSE: {mse_tree:.2f}")
print(f"Decision Tree Regression R²: {r2_tree:.2f}")


Linear Regression MSE: 0.65
Linear Regression R²: 0.71
Decision Tree Regression MSE: 0.03
Decision Tree Regression R²: 0.99


#Interpret the model's results and analyze the most


Key Evaluation Metrics:
Mean Squared Error (MSE): Measures the average squared difference between predicted and actual values. Lower values indicate better performance.
R-squared (R²): Measures the proportion of variance in the target variable explained by the model. R² closer to 1 indicates a good fit.

#influential features affecting restaurant ratings.

Features  affecting restaurant ratings are  'Has Table booking',
'Has Online delivery','Is delivering now','Switch to order menu','Price range','Rating color','Rating text'
